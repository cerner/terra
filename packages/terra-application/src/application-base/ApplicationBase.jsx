/* global TERRA_THEME_CONFIG */

import React, {
  useRef, useEffect, Suspense, useMemo, useState,
} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import ThemeProvider from 'terra-theme-provider';
import { ActiveBreakpointProvider } from 'terra-breakpoints';
import ThemeContextProvider from 'terra-theme-context/lib/ThemeContextProvider';
import { IntlProvider } from 'react-intl';
import i18nLoader from './private/i18nLoader';

import ApplicationErrorBoundary from '../application-error-boundary';
import { ApplicationIntlProvider } from '../application-intl';
import ApplicationLoadingOverlay, { ApplicationLoadingOverlayProvider } from '../application-loading-overlay';
import { ApplicationStatusOverlayProvider } from '../application-status-overlay';
import { NavigationPromptCheckpoint } from '../navigation-prompt';
import getBrowserLocale from './private/getBrowserLocale';
import useTestOverrides from './private/useTestOverrides';
import Logger from '../utils/logger';

import './private/initializeInert';
import './baseStyles';

import styles from './ApplicationBase.module.scss';

const cx = classNames.bind(styles);

const browserLocale = getBrowserLocale();

// We only need to retrieve the root theme and root theme name once for the life of the application.
const themeConfig = (typeof (TERRA_THEME_CONFIG) !== 'undefined') ? TERRA_THEME_CONFIG : undefined;
const rootThemeName = themeConfig?.theme ? themeConfig.theme : 'terra-default-theme';

const propTypes = {
  /**
   * The components to render within ApplicationBase.
   */
  children: PropTypes.node.isRequired,
  /**
   * The locale name to be used to load translated messages.
   * If the `locale` prop is not provided, the preferred language from the browser will be used.
   */
  locale: PropTypes.string,
  /**
   * The name of the theme to apply to the application using terra-theme-provider.
   */
  themeName: PropTypes.string,
  /**
   * By default, the elements rendered by ApplicationBase are fit to the Application's parent using 100% height.
   * If `fitToParentIsDisabled` is provided, the Application will render at its intrinsic content height and
   * potentially overflow its parent.
   */
  fitToParentIsDisabled: PropTypes.bool,
  /**
   * By default, NavigationPrompts rendered within ApplicationBase will cause the user to be prompted during
   * the window's beforeUnload event. If `unloadPromptIsDisabled` is provided, the user will **not** be prompted
   * before continuing with the unload event, even if NavigationPrompts are present.
   */
  unloadPromptIsDisabled: PropTypes.bool,
};

const ApplicationBase = ({
  locale, themeName, fitToParentIsDisabled, children, unloadPromptIsDisabled,
}) => {
  const registeredPromptsRef = useRef();
  const [messages, setMessages] = useState();

  useEffect(() => {
    if (unloadPromptIsDisabled) {
      return undefined;
    }

    function onBeforeUnload(event) {
      if (registeredPromptsRef.current && registeredPromptsRef.current.length) {
        event.preventDefault();

        // Chrome requires returnValue to be set to present the confirmation dialog
        event.returnValue = ''; // eslint-disable-line no-param-reassign

        // For this prompt, ApplicationBase is limited to browser-defaulted messaging.
        return '';
      }

      return undefined;
    }

    window.addEventListener('beforeunload', onBeforeUnload);

    return () => {
      window.removeEventListener('beforeunload', onBeforeUnload);
    };
  }, [unloadPromptIsDisabled, registeredPromptsRef]);

  const { localeOverride, themeOverride } = useTestOverrides(); // Allows us to test deployed applications in different locales.

  const finalLocale = localeOverride || locale || browserLocale;

  useEffect(() => {
    if (finalLocale !== undefined) {
      i18nLoader(finalLocale).then(translationsModule => setMessages(translationsModule.default)).catch((error) => {
        // eslint-disable-next-line no-console
        Logger.error(error);
        throw error;
      });
    }
  }, [finalLocale]);

  const theme = useMemo(() => ({
    // If the theme class name is undefined or an empty string, that indicates we have the root theme and should apply the root theme name.
    name: themeOverride || themeName || rootThemeName,
    className: themeOverride || themeName,
  }), [themeOverride, themeName]);

  if (messages === undefined) return null;

  return (
    <div data-terra-application-base className={cx('application-base', { fill: !fitToParentIsDisabled })}>
      <ThemeProvider themeName={themeName}>
        <ThemeContextProvider theme={theme}>
          <IntlProvider
            key={finalLocale}
            locale={finalLocale}
            messages={messages}
          >
            <ApplicationErrorBoundary>
              <ApplicationIntlProvider>
                <ActiveBreakpointProvider>
                  <NavigationPromptCheckpoint
                    onPromptChange={(registeredPrompts) => {
                      registeredPromptsRef.current = registeredPrompts;
                    }}
                  >
                    <ApplicationLoadingOverlayProvider>
                      <ApplicationStatusOverlayProvider>
                        <Suspense fallback={<ApplicationLoadingOverlay isOpen />}>
                          {children}
                        </Suspense>
                      </ApplicationStatusOverlayProvider>
                    </ApplicationLoadingOverlayProvider>
                  </NavigationPromptCheckpoint>
                </ActiveBreakpointProvider>
              </ApplicationIntlProvider>
            </ApplicationErrorBoundary>
          </IntlProvider>
        </ThemeContextProvider>
      </ThemeProvider>
    </div>
  );
};

ApplicationBase.propTypes = propTypes;

export default ApplicationBase;
