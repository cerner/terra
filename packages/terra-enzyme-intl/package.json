{
  "name": "@cerner/terra-enzyme-intl",
  "version": "4.6.1",
  "description": "Enzyme helpers for fortifying tests that depend on react-intl by decoupling the need for actual translations.",
  "author": "Cerner Corporation",
  "main": "lib/index.js",
  "engines": {
    "node": ">=10"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/cerner/terra.git",
    "directory": "packages/terra-enzyme-intl"
  },
  "bugs": {
    "url": "https://github.com/cerner/terra/issues"
  },
  "homepage": "https://github.com/cerner/terra/packages/terra-enzyme-intl#readme",
  "publishConfig": {
    "access": "public"
  },
  "license": "Apache-2.0",
  "keywords": [
    "Cerner",
    "enzyme",
    "i18n",
    "intl",
    "jest",
    "react",
    "react-intl",
    "Terra",
    "terra-enzyme-intl"
  ],
  "files": [
    "lib"
  ],
  "dependencies": {
    "prop-types": "^15.5.8"
  },
  "devDependencies": {
    "react-intl": "^5.25.1"
  },
  "peerDependencies": {
    "enzyme": "3.x",
    "react": "16.x",
    "react-intl": "^5.25.1"
  },
  "scripts": {
    "clean:install": "npm run clean && npm install",
    "compile": "babel --root-mode upward src --out-dir lib --copy-files",
    "jest": "jest --config ../../jest.config.js",
    "lint": "eslint --ext .js,.jsx . --ignore-path ../../.eslintignore",
    "precompile": "rm -rf lib",
    "test": "npm run lint && npm run jest"
  }
}
