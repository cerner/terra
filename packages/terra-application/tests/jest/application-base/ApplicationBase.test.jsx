import React from 'react';
import { mountWithIntl } from '@cerner/terra-enzyme-intl';
import ApplicationBase from '../../../src/application-base/ApplicationBase';

describe('ApplicationBase', () => {
  it('should render with minimal props', () => {
    const wrapper = mountWithIntl((
      <ApplicationBase locale="en">
        <div>content</div>
      </ApplicationBase>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with all props', () => {
    const wrapper = mountWithIntl((
      <ApplicationBase
        locale="en"
        unloadPromptIsDisabled
        themeName="test-theme"
      >
        <div>content</div>
      </ApplicationBase>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with the preferred browser locale', () => {
    const wrapper = mountWithIntl((
      <ApplicationBase>
        <div>content</div>
      </ApplicationBase>
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
