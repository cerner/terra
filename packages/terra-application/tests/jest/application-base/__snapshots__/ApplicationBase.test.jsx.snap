// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationBase should render with all props 1`] = `
<ApplicationBase
  locale="en"
  themeName="test-theme"
  unloadPromptIsDisabled={true}
/>
`;

exports[`ApplicationBase should render with minimal props 1`] = `
<ApplicationBase
  locale="en"
/>
`;

exports[`ApplicationBase should render with the preferred browser locale 1`] = `<ApplicationBase />`;
